### PR TITLE
Increase accuracy of ModelAdmin types

### PR DESCRIPTION
**Pull request by antonagestam**
Hi,

I started working on narrowing some types for `ModelAdmin` and reducing `Any`s. I got a bit ahead of myself and thought I'd make a PR before spending more time on this to verify you're willing to review/work with me once finalized.

I got a [strange error from pytest](https://gist.github.com/antonagestam/54430b1c02081fabd6bbae1eb9046368) on every failure, so I had a hard time working out which errors to expect, so that needs another iteration to fix the test cases. Usually I would try running a test case, see what error is yielded and verify that it makes sense, and then add its `# E: ...` comment. Do you have any ideas about what might be wrong with my setup?

I also ran into a few properties that aren't documented, I've added TODO comments on them and figured maybe we should remove them from the stubs entirely? What do you think?

Edit: I was also unsure about how to introduce "stub-only" types, will prefixing with underscore work, or will those types be exposed as well?

### CODE REVIEW

File: django-stubs/contrib/admin/options.pyi [hunk 1]
```diff
@@ -26,8 +30,11 @@ from django.db.models.fields import Field
 
 IS_POPUP_VAR: str
 TO_FIELD_VAR: str
-HORIZONTAL: Any
-VERTICAL: Any
+HORIZONTAL: Final[Literal[1]] = ...
```
Comment (sobolevn) [inline django-stubs/contrib/admin/options.pyi]: Isn't `Literal[1]` just enough?

File: django-stubs/contrib/admin/options.pyi [hunk 2]
```diff
@@ -26,8 +30,11 @@ from django.db.models.fields import Field
...
+VERTICAL: Final[Literal[2]] = ...
+
+# TODO: Is this ok? Is this the way to mark a type as stub-only?
+_Direction = Literal[1, 2]
```
Comment (sobolevn) [inline django-stubs/contrib/admin/options.pyi]: I guess `Union[Literal[1], Literal[2]]` is more readable.

File: django-stubs/contrib/admin/options.pyi [hunk 3]
```diff
@@ -37,24 +44,31 @@ class IncorrectLookupParameters(Exception): ...
...
+    radio_fields: Mapping[str, _Direction] = ...
+    prepopulated_fields: Mapping[str, Sequence[str]] = ...
+    formfield_overrides: Mapping[Type[Field], Mapping[str, Any]] = ...
+    readonly_fields: Sequence[str] = ...
+    ordering: Sequence[str] = ...
+    sortable_by: Sequence[str] = ...
     view_on_site: bool = ...
     show_full_result_count: bool = ...
-    checks_class: Any = ...
+    checks_class: Any = ...  # TODO: This seems to be undocumented, should we remove it?
```
Comment (sobolevn) [inline django-stubs/contrib/admin/options.pyi]: Let it be `Any` for now

File: django-stubs/contrib/admin/options.pyi [hunk 4]
```diff
@@ -93,35 +107,35 @@ class BaseModelAdmin:
     def has_module_permission(self, request: HttpRequest) -> bool: ...
 
 class ModelAdmin(BaseModelAdmin):
-    list_display: Sequence[Union[str, Callable]] = ...
+    list_display: Sequence[Union[str, Callable[[Model], Any]]] = ...
```
Comment (sobolevn) [inline django-stubs/contrib/admin/options.pyi:116]: Can we reduce `Any` here? Or is it hard to do?

File: django-stubs/contrib/admin/options.pyi [hunk 5]
```diff
@@ -93,35 +107,35 @@ class BaseModelAdmin:
...
-    action_form: Any = ...
+    add_form_template: str = ...
+    change_form_template: str = ...
+    change_list_template: str = ...
+    delete_confirmation_template: str = ...
+    delete_selected_confirmation_template: str = ...
+    object_history_template: str = ...
+    popup_response_template: str = ...
+    actions: Sequence[Callable[[ModelAdmin, HttpRequest, QuerySet], None]] = ...
+    action_form: Any = ...  # TODO: Undocumented, remove?
```
Comment (sobolevn) [inline django-stubs/contrib/admin/options.pyi]: Let's leave all undocumented feature for now.

**Review by sobolevn — COMMENTED** [review]
Looks great! Very big thanks! Much awesome!

File: django-stubs/contrib/admin/options.pyi (hunk 5 above)
Comment (antonagestam) [reply django-stubs/contrib/admin/options.pyi]: Gotcha! 👍

File: django-stubs/contrib/admin/options.pyi (hunk 4 above)
Comment (antonagestam) [reply django-stubs/contrib/admin/options.pyi:116]: I actually went for that at first, with a union of callables returning `str` or `bool` which are the two cases mentioned in the docs, but I now think this a case of the docs being vague. I think any type that is `str()`-able is a valid return type here, which is virtually any Python type, right? We could probably change to `object` here, not sure if that adds any value though.

File: django-stubs/contrib/admin/options.pyi (hunk 2 above)
Comment (antonagestam) [reply django-stubs/contrib/admin/options.pyi]: I'd disagree about the readability, but I'll update accordingly.

**Comment by antonagestam** [comment]
I managed to work around the strange pytest error by just putting the test case in a file and running mypy on it ... incredible how fast one gets used to nice tools!

File: django-stubs/contrib/admin/options.pyi [hunk 6]
```diff
@@ -26,8 +30,11 @@ from django.db.models.fields import Field
 
 IS_POPUP_VAR: str
 TO_FIELD_VAR: str
-HORIZONTAL: Any
-VERTICAL: Any
+HORIZONTAL: Literal[1] = ...
+VERTICAL: Literal[2] = ...
+
+# TODO: Is this ok? Is this the way to mark a type as stub-only?
```
Comment (antonagestam) [inline django-stubs/contrib/admin/options.pyi]: @sobolevn What's your opinion here? Do you use stub-only types in other places? Is this consider OK?

File: django-stubs/contrib/admin/options.pyi (hunk 4 above)
Comment (sobolevn) [reply django-stubs/contrib/admin/options.pyi:116]: Ok, let's leave `Any`. Thanks a lot for the explanation!

File: django-stubs/contrib/admin/options.pyi (hunk 6 above)
Comment (sobolevn) [reply django-stubs/contrib/admin/options.pyi]: Yes, this is fine.

**Review by sobolevn — APPROVED** [review]
Thanks 👍

File: django-stubs/contrib/admin/options.pyi [hunk 7]
```diff
@@ -227,9 +247,9 @@ class ModelAdmin(BaseModelAdmin):
     def history_view(self, request: HttpRequest, object_id: str, extra_context: None = ...) -> HttpResponse: ...
 
 class InlineModelAdmin(BaseModelAdmin):
-    model: Any = ...
-    fk_name: Any = ...
-    formset: Any = ...
+    model: Type[Model] = ...
+    fk_name: str = ...
+    formset: BaseFormSet = ...
```
Comment (sssbox) [inline django-stubs/contrib/admin/options.pyi:252]: I think this should be: `formset: Type[BaseFormSet] = ...`

Comment (sobolevn) [reply django-stubs/contrib/admin/options.pyi:252]: @sssbox, please submit a PR if you think that this is correct.

**Pull request merged**